File: training/llmviz_train/gpt.py
"""Minimal GPT reference implementation with activation recording.

This module is the golden source for the TypeScript port: every named
intermediate recorded here is compared against the TS re-implementation
at 1e-4 tolerance. Clarity therefore beats speed throughout — attention
is written out as explicit matmuls, no fused kernels.

Architecture (pre-norm residual stack):

    tokens -> tok_embedding + pos_embedding
           -> [x = x + attn_i(ln1_i(x)); x = x + mlp_i(ln2_i(x))] * n_layer
           -> ln_f -> lm_head (untied) -> logits

Norms are standard LayerNorm (eps 1e-5, with bias) — deliberately not
RMSNorm, because that is what GPTs use; the Mamba model uses RMSNorm.

CausalSelfAttention (head_dim = d_model / n_head):

    qkv_proj -> split (q, k, v), reshape to (B, n_head, T, head_dim)
    scores = q @ k^T / sqrt(head_dim), upper triangle masked to -inf
    weights = softmax(scores); out = out_proj(weights @ v)

MLP: fc (d_model -> mlp_ratio*d_model) -> GELU (exact erf) -> proj.
"""

import math

import torch
import torch.nn.functional as F
from torch import nn

from llmviz_train.mamba import _snapshot


class CausalSelfAttention(nn.Module):
    """Multi-head causal self-attention (no residual; the block adds it)."""

    def __init__(self, d_model: int, n_head: int):
        super().__init__()
        assert d_model % n_head == 0
        self.n_head = n_head
        self.head_dim = d_model // n_head

        self.qkv_proj = nn.Linear(d_model, 3 * d_model, bias=True)
        self.out_proj = nn.Linear(d_model, d_model, bias=True)

    def forward(
        self, x: torch.Tensor, record: dict[str, torch.Tensor] | None = None, prefix: str = ""
    ) -> torch.Tensor:
        """x: (B, T, d_model) -> (B, T, d_model).

        When ``record`` is a dict, named intermediates for batch item 0 are
        stored into it (detached cloned float32 CPU tensors) under ``prefix``.
        """
        B, T, d_model = x.shape

        def rec(name: str, tensor: torch.Tensor) -> None:
            if record is not None:
                record[prefix + name] = _snapshot(tensor)

        qkv = self.qkv_proj(x)  # (B, T, 3*d_model)
        q, k, v = qkv.chunk(3, dim=-1)  # each (B, T, d_model)
        # (B, T, d_model) -> (B, n_head, T, head_dim)
        q = q.view(B, T, self.n_head, self.head_dim).transpose(1, 2)
        k = k.view(B, T, self.n_head, self.head_dim).transpose(1, 2)
        v = v.view(B, T, self.n_head, self.head_dim).transpose(1, 2)
        rec("q", q[0])
        rec("k", k[0])
        rec("v", v[0])

        scores = q @ k.transpose(-2, -1) / math.sqrt(self.head_dim)  # (B, n_head, T, T)
        causal_mask = torch.triu(
            torch.ones(T, T, dtype=torch.bool, device=x.device), diagonal=1
        )
        scores = scores.masked_fill(causal_mask, float("-inf"))
        rec("scores", scores[0])  # post-mask, pre-softmax

        weights = F.softmax(scores, dim=-1)  # (B, n_head, T, T)
        rec("weights", weights[0])

        out = weights @ v  # (B, n_head, T, head_dim)
        out = out.transpose(1, 2).reshape(B, T, d_model)
        out = self.out_proj(out)
        rec("out", out[0])
        return out


class MLP(nn.Module):
    """fc -> GELU (exact erf, nn.GELU default) -> proj."""

    def __init__(self, d_model: int, mlp_ratio: int):
        super().__init__()
        d_mlp = mlp_ratio * d_model
        self.fc = nn.Linear(d_model, d_mlp, bias=True)
        self.act = nn.GELU()
        self.proj = nn.Linear(d_mlp, d_model, bias=True)

    def forward(
        self, x: torch.Tensor, record: dict[str, torch.Tensor] | None = None, prefix: str = ""
    ) -> torch.Tensor:
        """x: (B, T, d_model) -> (B, T, d_model), recording like the attn."""

        def rec(name: str, tensor: torch.Tensor) -> None:
            if record is not None:
                record[prefix + name] = _snapshot(tensor)

        pre = self.fc(x)  # (B, T, d_mlp)
        rec("fc", pre[0])
        post = self.act(pre)
        rec("act", post[0])
        out = self.proj(post)
        rec("proj", out[0])
        return out


class GPT(nn.Module):
    """Minimal GPT language model over a small token vocabulary."""

    def __init__(self, cfg: dict, vocab_size: int, max_seq_len: int = 32):
        super().__init__()
        n_layer = cfg["n_layer"]
        d_model = cfg["d_model"]
        self.max_seq_len = max_seq_len

        self.tok_embedding = nn.Embedding(vocab_size, d_model)
        self.pos_embedding = nn.Embedding(max_seq_len, d_model)
        self.ln1s = nn.ModuleList(nn.LayerNorm(d_model) for _ in range(n_layer))
        self.attns = nn.ModuleList(
            CausalSelfAttention(d_model, cfg["n_head"]) for _ in range(n_layer)
        )
        self.ln2s = nn.ModuleList(nn.LayerNorm(d_model) for _ in range(n_layer))
        self.mlps = nn.ModuleList(MLP(d_model, cfg["mlp_ratio"]) for _ in range(n_layer))
        self.ln_f = nn.LayerNorm(d_model)
        self.lm_head = nn.Linear(d_model, vocab_size, bias=False)  # untied

    def forward(
        self, tokens: torch.Tensor, record: bool = False
    ) -> torch.Tensor | tuple[torch.Tensor, dict[str, torch.Tensor]]:
        """tokens: (B, T) long -> logits (B, T, vocab).

        With ``record=True``, also returns a dict of named activations for
        batch item 0 (detached cloned float32 CPU tensors).
        """
        T = tokens.shape[1]
        assert T <= self.max_seq_len, f"T={T} exceeds max_seq_len={self.max_seq_len}"
        acts: dict[str, torch.Tensor] | None = {} if record else None

        def rec(name: str, tensor: torch.Tensor) -> None:
            if acts is not None:
                acts[name] = _snapshot(tensor)

        positions = torch.arange(T, device=tokens.device)
        x = self.tok_embedding(tokens) + self.pos_embedding(positions)
        rec("embed.out", x[0])

        for i in range(len(self.attns)):
            normed = self.ln1s[i](x)
            rec(f"layer{i}.ln1.out", normed[0])
            x = x + self.attns[i](normed, record=acts, prefix=f"layer{i}.attn.")
            normed = self.ln2s[i](x)
            rec(f"layer{i}.ln2.out", normed[0])
            x = x + self.mlps[i](normed, record=acts, prefix=f"layer{i}.mlp.")

        x = self.ln_f(x)
        rec("final_norm.out", x[0])

        logits = self.lm_head(x)
        rec("head.logits", logits[0])

        if acts is not None:
            return logits, acts
        return logits

File: src/compute/gpt.ts
/**
 * TS port of the GPT forward pass in training/llmviz_train/gpt.py,
 * operation-for-operation, recording the same activation names. Gated by
 * gpt.golden.test.ts against PyTorch goldens at 1e-4.
 *
 * Architecture (pre-norm residual stack):
 *
 *     tokens -> tok_embedding + pos_embedding
 *            -> [x = x + attn_i(ln1_i(x)); x = x + mlp_i(ln2_i(x))] * n_layer
 *            -> ln_f -> lm_head (untied) -> logits
 *
 * CausalSelfAttention (head_dim = d_model / n_head):
 *
 *     qkv_proj -> split (q, k, v)               [torch.chunk order]
 *     per head: scores = q @ k^T / sqrt(head_dim),
 *               strict upper triangle masked to -inf BEFORE softmax,
 *               out_h = softmax(scores) @ v
 *     out = out_proj(concat heads)
 *
 * The batch dim of gpt.py is dropped (the goldens record batch item 0);
 * heads are processed as 2D (T, head_dim) / (T, T) slices and re-stacked
 * into (n_head, T, ·) tensors only for recording, matching the reference's
 * (B, n_head, T, ·) layout at B index 0.
 *
 * MLP: fc -> GELU (exact erf) -> proj.
 */

import { T } from "./tensor";
import { add, embedding, gelu, layernorm, linear, matmul, scale, softmax, transpose2d } from "./ops";
import { getTensor, type Manifest } from "./loader";
import type { Recorder } from "./recorder";
import { dimsFromManifest, sliceCols } from "./util";

/** GPT dims as exported in manifest.json (head_dim = d_model / n_head). */
export interface GptDims {
  n_layer: number;
  d_model: number;
  n_head: number;
  mlp_ratio: number;
  vocab_size: number;
  head_dim: number;
  max_seq_len: number;
}

const GPT_DIM_KEYS = [
  "n_layer",
  "d_model",
  "n_head",
  "mlp_ratio",
  "vocab_size",
  "head_dim",
  "max_seq_len",
] as const;

/** Validated GptDims from a manifest (throws on wrong arch / missing keys). */
export function gptDimsFrom(manifest: Manifest): GptDims {
  const dims = dimsFromManifest(manifest, "gpt", GPT_DIM_KEYS);
  if (dims.n_head * dims.head_dim !== dims.d_model) {
    throw new Error(
      `gpt dims inconsistent: n_head ${dims.n_head} * head_dim ${dims.head_dim} ` +
        `!= d_model ${dims.d_model}`,
    );
  }
  return dims;
}

/** Stack n same-shaped 2D tensors (r,c) into one (n,r,c) tensor (for recording). */
function stack(slices: T[]): T {
  const [r, c] = slices[0].shape;
  const out = T.zeros([slices.length, r, c]);
  for (let i = 0; i < slices.length; i++) {
    out.data.set(slices[i].data, i * r * c);
  }
  return out;
}

/** Minimal attention dims (a structural subset of GptDims and MoeDims). */
export interface AttnDims {
  d_model: number;
  n_head: number;
  head_dim: number;
}

/**
 * One CausalSelfAttention (gpt.py CausalSelfAttention.forward), no
 * residual — the caller adds it. x: (T, d_model) -> (T, d_model).
 * Records q/k/v/scores/weights/out under `prefix`. Exported so MoE (whose
 * attention is identical, same `attns.{i}.*` weight names) can reuse it.
 */
export function causalSelfAttention(
  weights: Map<string, T>,
  dims: AttnDims,
  layer: number,
  x: T,
  rec: Recorder | undefined,
): T {
  const { d_model: dModel, n_head: nHead, head_dim: headDim } = dims;
  const steps = x.shape[0];
  const prefix = `layer${layer}.attn.`;
  const w = (suffix: string): T => getTensor(weights, `attns.${layer}.${suffix}`);

  // qkv_proj -> split (q, k, v) in torch.chunk(3, dim=-1) order.
  const qkv = linear(x, w("qkv_proj.weight"), w("qkv_proj.bias")); // (T, 3*d_model)
  const q = sliceCols(qkv, 0, dModel);
  const k = sliceCols(qkv, dModel, 2 * dModel);
  const v = sliceCols(qkv, 2 * dModel, 3 * dModel);

  // view(T, n_head, head_dim).transpose(0, 1): head h owns columns
  // [h*head_dim, (h+1)*head_dim) of the (T, d_model) matrix.
  const qh: T[] = [];
  const kh: T[] = [];
  const vh: T[] = [];
  for (let h = 0; h < nHead; h++) {
    qh.push(sliceCols(q, h * headDim, (h + 1) * headDim));
    kh.push(sliceCols(k, h * headDim, (h + 1) * headDim));
    vh.push(sliceCols(v, h * headDim, (h + 1) * headDim));
  }
  rec?.record(`${prefix}q`, stack(qh)); // (n_head, T, head_dim)
  rec?.record(`${prefix}k`, stack(kh));
  rec?.record(`${prefix}v`, stack(vh));

  // Per head: scores = q @ k^T / sqrt(head_dim), strict upper triangle
  // masked to -inf (post-mask recorded), softmax, then weights @ v.
  const scoresH: T[] = [];
  const weightsH: T[] = [];
  const outH: T[] = [];
  for (let h = 0; h < nHead; h++) {
    const scores = scale(matmul(qh[h], transpose2d(kh[h])), 1 / Math.sqrt(headDim)); // (T, T)
    for (let i = 0; i < steps; i++) {
      for (let j = i + 1; j < steps; j++) {
        scores.data[i * steps + j] = -Infinity;
      }
    }
    scoresH.push(scores);
    const wts = softmax(scores); // (T, T)
    weightsH.push(wts);
    outH.push(matmul(wts, vh[h])); // (T, head_dim)
  }
  rec?.record(`${prefix}scores`, stack(scoresH)); // (n_head, T, T), post-mask
  rec?.record(`${prefix}weights`, stack(weightsH));

  // transpose(1, 2).reshape(T, d_model): re-interleave heads column-wise.
  const merged = T.zeros([steps, dModel]);
  for (let h = 0; h < nHead; h++) {
    for (let t = 0; t < steps; t++) {
      merged.data.set(outH[h].data.subarray(t * headDim, (t + 1) * headDim), t * dModel + h * headDim);
    }
  }

  const out = linear(merged, w("out_proj.weight"), w("out_proj.bias"));
  rec?.record(`${prefix}out`, out);
  return out;
}

/**
 * One MLP (gpt.py MLP.forward): fc -> GELU (exact erf) -> proj, no
 * residual. x: (T, d_model) -> (T, d_model). Records under `prefix`.
 */
function mlp(
  weights: Map<string, T>,
  layer: number,
  x: T,
  rec: Recorder | undefined,
): T {
  const prefix = `layer${layer}.mlp.`;
  const w = (suffix: string): T => getTensor(weights, `mlps.${layer}.${suffix}`);

  const pre = linear(x, w("fc.weight"), w("fc.bias")); // (T, d_mlp)
  rec?.record(`${prefix}fc`, pre);
  const post = gelu(pre);
  rec?.record(`${prefix}act`, post);
  const out = linear(post, w("proj.weight"), w("proj.bias"));
  rec?.record(`${prefix}proj`, out);
  return out;
}

/**
 * Full forward pass: tokens -> logits (T, vocab), recording the same
 * activation names as gpt.py when `rec` is given.
 */
export function gptForward(
  weights: Map<string, T>,
  dims: GptDims,
  tokens: number[],
  rec?: Recorder,
): T {
  if (tokens.length > dims.max_seq_len) {
    throw new Error(`T=${tokens.length} exceeds max_seq_len=${dims.max_seq_len}`);
  }

  const positions = tokens.map((_, i) => i);
  let x = add(
    embedding(tokens, getTensor(weights, "tok_embedding.weight")),
    embedding(positions, getTensor(weights, "pos_embedding.weight")),
  ); // (T, d_model)
  rec?.record("embed.out", x);

  for (let i = 0; i < dims.n_layer; i++) {
    let normed = layernorm(
      x,
      getTensor(weights, `ln1s.${i}.weight`),
      getTensor(weights, `ln1s.${i}.bias`),
    );
    rec?.record(`layer${i}.ln1.out`, normed);
    x = add(x, causalSelfAttention(weights, dims, i, normed, rec));

    normed = layernorm(
      x,
      getTensor(weights, `ln2s.${i}.weight`),
      getTensor(weights, `ln2s.${i}.bias`),
    );
    rec?.record(`layer${i}.ln2.out`, normed);
    x = add(x, mlp(weights, i, normed, rec));
  }

  x = layernorm(x, getTensor(weights, "ln_f.weight"), getTensor(weights, "ln_f.bias"));
  rec?.record("final_norm.out", x);

  const logits = linear(x, getTensor(weights, "lm_head.weight"));
  rec?.record("head.logits", logits);
  return logits;
}
